### DISCUSSION TITLE
BlueRetro HW2 v1.8 BETA

### DISCUSSION THREAD

**Discussion by darthcloud** [darthcloud/BlueRetro#415] [Show and tell]
# This is a BETA FW and documentation is incomplete, it's only intended for people with high experience in electronic design.

v1.8 firmware introduce a new hardware specification called HW2 which bring the new port detection feature.

Any existing BlueRetro hardware now gets retroactively named HW1 spec.

From now on, two firmware zip archive will be released for new firmware, one for HW1 and one for HW2.

The main benefits of using BlueRetro HW2 is that internal install is officially supported and the port detection feature is used to avoid bus conflicts when wired controllers are used. Another benefits is power management from BT devices.

HW2 firmware can also be used for external adapter. The benefit here is that for multi connector adapter, unused cord are detected and disable. (in HW1 all populated cable are required to be used to prevent adapter instability)

The web OTA update page got a guard that will prevent flashing HW2 FW on HW1 hardware and vice-versa.
If you update and existing DIY HW1 adapter to HW2 spec you will need to first factory reset the adapter and then reflash firmware via USB.

Documentation is still not done, so you need to reference the high level specification pages:
Internal install:
https://github.com/darthcloud/BlueRetro/wiki/BlueRetro-HW2-Internal-Install-Specification

External adapter:
https://github.com/darthcloud/BlueRetro/wiki/BlueRetro-HW2-External-Specification

~~Try to use the beta web config aswell, no functional difference but a lot of code rework was done:~~
Current build wont work with beta web, use blueretro.io
https://darthcloud.github.io/BlueRetroWebCfg_Beta/

# Only GameCube FW was tested, other system spec is subject to change for final release.

For HW2
~~v1.8-rc2_hw2.zip~~

For HW1
~~v1.8-rc2_hw1.zip~~

**Comment by Alf-Tupper** [comment]
Hi @darthcloud,

Absolutely love what you're doing! Was looking for a wireless solution for my PS2 and was trying to source a UnoRetro Adaptor, and had no joy... so decided to have a go with the ESP32 myself.

I installed internally on my PS2 with V1.7.3. Also fitted separate external buttons and a usb extension, all works very well. Need to press the ESP32 reset after a power off, but I'm ok with that.  

I spotted the V1.8, so I thought id give it a go, flashed fine, no issues. My PS4 controller paired first time, but as player 3 or 4 if I connect my other controller. Is there something I need to update or should I just go back to V1.7.3 and wait for the final version?

Cheers,

Dean

**Comment by darthcloud** [comment]
Here v1.8-rc4 which add support for GameID based config.

You need to use the beta web-config link to see the new options:
https://darthcloud.github.io/BlueRetroWebCfg_Beta/

# Limitation
First BlueRetro got a limit of 16 files on it's file system. This isn't related to space on the partition but rather due to the heavy usage of RAM the filesystem consume when the maximum files count is high.

BlueRetro got 4 system files to begin with so this leave space for a **maximum of 12 extra configs for GameID.**

# Managing the file system
I added a new page to allow removing the configs once you reach the limit:
https://darthcloud.github.io/BlueRetroWebCfg_Beta/files.html

By default their is only 4 default files. Those files are re-created automatically at boot if missing.
config.bin is the default config.
![chrome_06QO6uZOlm](https://user-images.githubusercontent.com/3744056/201498019-762ec777-ad7d-4df7-af4d-9579e05535c0.png)

Once you got a few config it will look like this:
![ApplicationFrameHost_bCuuwf0RMM](https://user-images.githubusercontent.com/3744056/201498088-0338a785-27b1-4416-ae4e-7dea13f2a867.png)

The file name will be base on the RAW GameId value, place the cursor over the file name to make an info box appear with the actual Game name.

# How to switch a Game to it's own config

1. Using a loader that support GameId (Xstation, EverDrive, Swiss-GC) start the game than disconnect every bluetooth device from BlueRetro.
2. Connect to either the Advance config or Presets page at https://darthcloud.github.io/BlueRetroWebCfg_Beta/
3. Once connected the current active Game name will be displayed:
![chrome_bS6vtzBwQa](https://user-images.githubusercontent.com/3744056/201498278-1e794a2f-b114-4b59-ad26-562cf50d8af2.png)
4. Click on Switch to GameID button
5. Now every times this game is active this config will be loaded automatically and any change will be saved in that file.
![chrome_DeLmco7xT6](https://user-images.githubusercontent.com/3744056/201498349-d0902597-3152-4b82-b2d9-79ca27ccb5fa.png)

# Per console type config
For the universal version of BlueRetro the GameID framework can be reused to save different default config for different system.
Simply boot BlueRetro on the system you want to set a new default and follow the same instruction as above.

The game name will be undefined but the GameID will be the console name.
![chrome_HsiXc9wf2m](https://user-images.githubusercontent.com/3744056/201498470-ce737ae6-6606-49b2-ba79-470051d32610.png)

# Firmware
~~v1.8-rc4_hw1.zip~~
~~v1.8-rc4_hw2.zip~~

**Comment by darthcloud** [comment]
RC5 fix again some issues related to malloc. 

~~v1.8-rc5_hw1.zip~~
~~v1.8-rc5_hw2.zip~~
